templates/todo/projects/java-mongo-aca/.repo/bicep/infra/main.bicep

```diff
@@ -93,6 +93,7 @@ module api '../../../../../common/infra/bicep/app/api-container-app.bicep' = {
     containerAppsEnvironmentName: containerApps.outputs.environmentName
     containerRegistryName: containerApps.outputs.registryName
     keyVaultName: keyVault.outputs.name
+    corsACAURL: join(['https://${abbrs.appContainerApps}api-${resourceToken}.',containerApps.outputs.defaultDomain],'')
```

Comment:
You'll need to use this as an output not construct the name here again... because you'll have name drift if you do it like this

Reply:
that goes for all locations that you are constructing this URL

Reply:
We cannot use an output for this because it creates a circular dependency when doing so (this would create a dependency from the `web` module to the `api` module, and the `api` module already has a dependency on the `web` module, and bicep will not allow the circular dependency.

We need a way in this file to determine what the endpoint that the web server will have so that we can set it as an environment variable in the settings of the API server so that the implementation of the API server can include it in the list of domains it allows access to via CORS.

If we don't want to solve the problem this way, I think the other option we have for decoupling would be to spin up an AppConfiguration instance and have that store properties.  Then use the AppConfiguration SDK to fetch these properties instead of reading them from environment variables.

Reply:
BTW, @hemarina - surprised to see that we have `api-` not `web` here?  I would have expected that we need to be giving the URL of the host running the website to the API service, I think this will cause us to give CORS permissions to the wrong domain.

Reply:
it should be `web`.  But it is very fragile, as this is a hardcoded value,... which is broken as soon as the service name is changed

Reply:
> it should be web. But it is very fragile, as this is a hardcoded value,... which is broken as soon as the service name is changed.

The fundamental problem we have here is that the `api` service needs to know the endpoint of the `web` service, so it can include it in the list of domains it allows CORS access to.  The `web` service needs to know about the endpoint of the `api` service so that it knows what host to connect to for API requests.

Since we can't take a direct reference on the outputs, because that would create a circular dependency, we have two options:

1. Be clever and compute whatever name the endpoint is going to get and pass that as a parameter to one of these two components, to break the cycle (that's what this change does by trying to compute the endpoint name that the container will be assigned).  Looking at it more, I see that we allow the container name to be changed like this:

```bicep
name: !empty(apiContainerAppName) ? apiContainerAppName : '${abbrs.appContainerApps}api-${resourceToken}'
```

And in the case where you set `apiContainerAppName`, the thing we compute would be wrong.  We should instead probably define a local variable called `apiContainerAppNameOrDefault` which gets the value of this expression and then use `apiContainerAppNameOrDefault` for the name parameter and in the interpolation of `corsAcaUrl` (i.e. `'https://${apiContainerAppNameOrDefault}.${containerApps.outputs.defaultDomain}'`.

Alternately we can break the cycle by using a layer of indirection, via Azure App Configuration.  In that would I expect we would do the following:

1. Modify `main.bicep` to create an [`Microsoft.AppConfiguration/configurationStores`](https://learn.microsoft.com/en-us/azure/templates/microsoft.appconfiguration/2020-07-01-preview/configurationstores?pivots=deployment-language-bicep)

2. Modify `main.bicep` to create entries in the configuration store for the api endpoint and the web endpoint (we would create an [`Microsoft.AppConfiguration/configurationStores/keyValues`](https://learn.microsoft.com/en-us/azure/templates/microsoft.appconfiguration/configurationstores/keyvalues?pivots=deployment-language-bicep) for each)

3. Modify the TODO sample app to read these values from App Configuration instead of process environment variables.  As a policy decision, it might do this in cases where the values are not set in the environment, so we have a better local dev story (or we can just say you take a dependency on the provisioned app config resource, similar to Key Vault).

The api and web services would need references to the configuration store, but the store can be provisioned independently of the keys, and so we would no longer have this circular dependency.

Reply:
What if we make `setting CORS` to be a `deploy` task.
If we think about it, CORS is not required until a server (api) is deployed to Azure.
So, `deploy` could do the same thing it does for web app (when it sets the api-url as env var), instead of expecting `provision` to set the CORS-endpoint env var....
Deploy would parse any `WEB_URL` within azd environment and get the hostname from it, and then it would set it as env var for the server app. Finally, the server app uses the value to set CORS :).

Reply:
> Deploy would parse any WEB_URL within azd environment and get the hostname from it, and then it would set it as env var for the server app. Finally, the server app uses the value to set CORS :).

I think the question becomes how do we intend to do this?  Changing the app settings requires mutating properties of the deployed resource (in this case the container app resource).  We would need to make sure that running `azd provision` after running `azd deploy` did not reset these settings, and I'm not sure how exactly we'd end up doing that.  I'm also not sure how we would express what we wanted `azd` to do here and how it would be configured.

I think this runs into the same problems we had with our initial container support, where we'd change parts of the resource after doing a provision (the container image to use) and then needed to do work to make sure `provision` didn't regress things.

I'm worried that teaching `azd` something specific about CORS or trying to build a set of features that let you do a limited form of service discovery.

Thinking on it more, it sounds like you are proposing that we change how the API service is packaged such that we burn in the name of the endpoint into the binary instead of picking it up from some config system (or we do pick it up from config, but it's config that is generated at application deploy time).  I think I understand how that could work - but it feels like that's moving us further away from where we want to be.  The application really should be agnostic to this configuration and pull it in from and external source (either environment variables or some centralized config store, or a service discovery system) instead of the deployed application package being tied to a specific set of endpoints.

Reply:
> it sounds like you are proposing that we change how the API service is packaged such that we burn in the name of the endpoint into the binary instead of picking it up from some config system

Yes, just like we are currently doing with the `web-app`.
I am thinking that we can create some abstraction or setting within the service definition (maybe in azure.yaml, or somewhere) to define what env-var are relevant for the service and which `azd` should burn as app-config or at least as a dotenv file that is packed next to the app artifact that is deployed.

Reply:
when ever i have bicep circular dependency I break the offending section to a separate module.  Is that possible in this case?